### PR TITLE
Sort name ascii2

### LINKED ISSUE #1500: sorting PacBio bam by name might not be well supported.
```bash
$ samtools --version
samtools 1.13
Using htslib 1.13

$ uname -sr
Linux 5.4.0-1052-gcp

$ uname -m
x86_64

$ gcc --version
gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
```

It looks like `samtools sort -n` might not work well with PacBio reads?

Below is the bam after the sort by name (Picard `MarkDuplicates` complained).
For example, reads 
```
000a9e2b-787a-4241-9db9-a8600bd3c497
000a43da-5840-421d-8b5a-3f83b5a048dc
```

```bash
$ samview -h HG00732.1-p.name-sorted.bam | grep -v "@SQ" | less -S
@HD     VN:1.6  SO:queryname
@RG     ID:6061-SL-0002a        SM:HG00732      PL:ONT  PU:     DT:
@RG     ID:6061-SL-0002b        SM:HG00732      PL:ONT  PU:     DT:
@RG     ID:6061-SL-0002c        SM:HG00732      PL:ONT  PU:     DT:
@RG     ID:6061-SL-0002d        SM:HG00732      PL:ONT  PU:     DT:
@PG     ID:minimap2     PN:minimap2     VN:2.17-r941    CL:minimap2 -ayYL --MD -x map-ont -R @RG\tID:6061-SL-0002a\tSM:HG00732\tPL:ON
@PG     ID:samtools     PN:samtools     PP:minimap2     VN:1.13 CL:samtools view -bhX -M -@ 1 --verbosity=8 --write-index -o HG00732.
@PG     ID:samtools.1   PN:samtools     PP:samtools     VN:1.13 CL:samtools sort -n -@ 1 -o HG00732.1-p.name-sorted.bam HG00732.1-p.b
@PG     ID:samtools.2   PN:samtools     PP:samtools.1   VN:1.13 CL:samtools view -h HG00732.1-p.name-sorted.bam
0000a0eb-63fc-43ee-b1dd-0cc129732b58    0       chr1    99724923        60      22S12M1D7M1D8M2D9M1I13M1D11M1I2M4D15M2I6M4D31M2I30M1I
0000acac-3c6e-41da-a4e5-cac6c28d34a1    16      chr1    97918692        60      12S48M1I13M4I11M1D19M2D73M1D6M1I16M2D15M1D8M1D26M2I18
0000b5ae-cb42-4fd3-aaf7-63e9cfe65431    16      chr1    78294524        60      10M1D8M1D4M1D20M1D4M1D14M1D6M1I35M1D13M1D8M1D10M3I11M
0000bd07-198b-4e8a-b708-9c54c2225372    16      chr1    111129409       60      9S24M1I2M1D16M1D33M3I127M1D12M1I109M2D73M1D29M1D52M1D
0000e4ce-0654-4ee8-8680-92d2e18032ad    0       chr1    121774415       1       5S18M1D16M1I26M1D4M1D24M3D11M2I17M1D4M1D42M1I39M1I11M
0000e4ce-0654-4ee8-8680-92d2e18032ad    256     chr1    122744575       0       5S18M1D16M1I26M1D4M1D24M3D11M2I17M1D4M1D42M1I39M1I4M1
0000fe75-1232-44da-8b50-24eed7c7eaad    272     chr1    122627141       0       25S7M1I3M1D8M1D3M1D77M2I63M1I109M1I17M1D37M3I7M1D105M
0000fe75-1232-44da-8b50-24eed7c7eaad    272     chr1    122812077       0       17S15M1I3M1D8M2D81M1I1M1I61M1I109M1I17M1D37M3I7M1D105
0000fe75-1232-44da-8b50-24eed7c7eaad    272     chr1    123358129       0       19S13M1I3M1D8M2D81M1I1M1I26M4I34M1D2M2I103M1I17M1D38M
000a0b6e-283d-4e03-ba4c-72c369faf73f    0       chr1    107618854       60      7S51M1I50M1D18M1I18M1I36M1D30M5D34M1I33M1D81M1I40M2D5
000a3fc5-c97f-4f91-9cbf-271eb06d04d1    2064    chr1    78551746        60      14S45M1D44M2D35M1D6M1D16M1I15M2I37M2D71M3I22M1D21M1D7
000a4b7f-ddee-47b0-a847-61d88847a6be    16      chr1    117486660       60      12S36M1D6M2D52M1I20M1I192M1I10M3D82M3D3M1D18M2I29M1I1
000a5d34-9c0a-4f31-b38f-2c22ce84e527    0       chr1    105940422       60      30S26M5D2M1D5M1D19M1D4M2I41M6D21M1D64M1D67M1D28M1I10M
000a8bd9-9324-4003-8f81-2c2b12505c20    0       chr1    51438347        60      25S55M3D29M1D64M3D37M1I9M1D36M1D58M1D48M1I44M1D28M1D6
000a9e2b-787a-4241-9db9-a8600bd3c497    16      chr1    115360502       60      297S15M1D1M1D7M1D6M2D9M1I2M2D5M1D2M2D9M1I9M1I3M2D3M1I
000a43da-5840-421d-8b5a-3f83b5a048dc    0       chr1    121762714       1       20S90M1I41M2D20M3I13M1I2M1I73M1D44M1D27M1I77M1I7M2I43
000a43da-5840-421d-8b5a-3f83b5a048dc    256     chr1    121767124       0       17S93M1I32M1I3M3D25M3I13M1I2M1I73M1D44M1D27M1I77M1I7M
000a060f-4431-4631-aba4-93d76dfce08a    16      chr1    40336837        60      182S12M2I38M2I30M3I11M2I1M3I16M1D6M4I5M1I5M2I14M7I53M
000a552f-aba1-4403-b97c-a75f7cbed0e0    16      chr1    58671267        60      15S17M1I31M2D13M1I9M2I18M1I20M1D18M1D19M1D8M2D1M1I115
000a890b-c39f-4378-b2b8-a4ce33cad2ea    0       chr1    47700418        60      697S5M2I34M1D6M1I14M1D3M1D14M1I7M2I13M2I18M1I113M1D35
000a940a-3c0d-4402-962b-b5f04b3b0201    256     chr1    121197146       0       4S29M1I30M2I21M1D14M2D70M1D33M1I14M1D28M1I28M1D65M1D1
000a5371-5612-4552-8c57-731fafeb5a35    16      chr1    53251904        60      15373S52M1I27M2D32M2D23M1D103M1I4M1D47M1I3M1D6M3D110M
000a6930-f727-4607-a268-374c6e5f5dba    272     chr1    121269349       0       10S7M1D48M1D4M1D31M2D3M1D6M1I59M2D16M1D20M2D22M2D25M1
000aa0ea-0d57-44ba-abaa-36785ea9ee0e    16      chr1    62162874        60      14S74M1D5M5D62M1I38M3I8M1I145M3D46M1I33M1D7M2I9M2I46M
```

I am aware that sorting by read name means natural ordering, from the `man` page:
>Names are compared so as to give a “natural” ordering — i.e. sections consisting of digits are compared numerically while all other sections are compared based on their binary representation.

But this doesn't seem to comply with natural ordering.
Did I do something wrong? Or misunderstood something?

Thanks!
Steve

[test.sam.txt](https://github.com/samtools/samtools/files/7164709/test.sam.txt)

### PR DESCRIPTION
**Pull request by jkbonfield**
Add samtools sort -N / samtools merge -N option.

These handle an alternative form of read-name sort order that is purely ASCII instead of the "natural" alphanumeric sort.

Natural sort order will place "a12" after "a3", as it computes 12 > 3 rather than "1" < "3" (ASCII).  However it also means hexadecimal values such as "a12b" > "a3bc", and "a09b" > "a7bc".  Typically any read name generation system using leading fixed-width columns with leading zeros to pad is going to be safer using traditional ASCII sort orders.

Note while writing this my initial implementation was to add QueryNameASCII to the SamOrder typedef, but it turns out the sheer level of code duplication going on here means doing that yielding 81 changed lines, vs 15 for this.  I don't like using a global variable, but it's already using one due to the lacklustre ksort API which can't pass in a client state argument. (Being poorly modelled on qsort I assume.)  Hence a sibling global is a trivial change and less error prone.

Fixes #1500

Also corrected a poorly written (and incomplete) error checking on compatibility of `--write-index` with non-coordinate sort orders.

### CODE REVIEW

**Comment by jkbonfield** [comment]
I forgot to add a comment about subsort.

In theory we could amend the output to include sort-order as `queryname:natural` and `queryname:ASCII` (or similar terms).  However it doesn't really solve much.  If we are trying to merge 2 files, one claiming `queryname:natural` and the other just `queryname` then we still don't know what that means.  What the naked queryname an old samtools (so "natural") or picard (ASCII)?

Plus we have [code already](https://github.com/samtools/samtools/blob/develop/bam_markdup.c#L1483) that explicitly looks for "queryname".  We can fix the things we know about, but what about the ones we don't?  We'd run the risk of breaking all sorts of pipelines by changing `queryname` to `queryname:natural`, meaning the only safe thing is for the new `-N` and have an implicit assumption on non-sub-sort ordering to be natural (only it's not, as it's program dependent).

In short, it's a can of worms with no obvious correct solution.  Hence for now I dodged it.  Suggestions welcomed, or we can delay until a new PR.

**Comment by daviesrob** [comment]
On the subsort information - the [SAM specification suggests](https://samtools.github.io/hts-specs/SAMv1.pdf#subsubsection.1.3.1) this should go in the `SS` tag, not `SO`.  In fact `SO` only allows the values `unknown`, `unsorted`, `queryname` and `coordinate`.  Section 1.3.1 defines the terms `lexicographical` and `natural` for the two different string orderings.  Unfortunately it's not entirely clear how they should be applied in the `queryname` case as they're not strictly a sub-sort, but I'd guess the intention was `SS:queryname:lexicographical` or `SS:queryname:natural`?

**Comment by SHuang-Broad** [comment]
@jkbonfield , I just realized that some other tools might also need to be updated (including their documentation).

For example, `samtools merge` currently supports `-n`, so with this PR, `merge` probably should be updated accordingly.

**Comment by jkbonfield** [comment]
Good point - I was clearly misremembering SO vs SS.  Yes we could safely add SS, if we can decide what it should be.

https://github.com/samtools/hts-specs/pull/361 has appropriate advice to follow here.  I'll amend this PR.

File: doc/samtools-sort.1 [hunk 1]
```diff
@@ -236,11 +254,17 @@ Character tags (type A) are compared by binary character value.
...
-consisting of digits are compared numerically while all other sections are
-compared based on their binary representation.  This means \*(lqa1\*(rq will
-come before \*(lqb1\*(rq and \*(lqa9\*(rq will come before \*(lqa10\*(rq.
+When the \fB-n\fP or \fB-N\fP option is present, records are sorted by
+name.  Historically samtools has used a \*(lqnatural\*(rq ordering
+\(em i.e. sections consisting of digits are compared numerically while
+all other sections are compared based on their binary representation.
+This means \*(lqa1\*(rq will come before \*(lqb1\*(rq and \*(lqa9\*(rq
+will come before \*(lqa10\*(rq.  However this alpha-numeric sort can
+be confused by runs of hexademical digits.  The newer \fB-N\fP
```
Comment (daviesrob) [inline doc/samtools-sort.1:263]:
```suggestion
be confused by runs of hexadecimal digits.  The newer \fB-N\fP
```
